### PR TITLE
tools/netqtop: fix rx queue id

### PR DESCRIPTION
**Pull request by pizhenwei**
Net device driver uses 'skb_record_rx_queue' to mark rx queue
mapping during receiving a packet. Also need to call
'skb_get_rx_queue' to get rx queue id.

Signed-off-by: zhenwei pi <pizhenwei@bytedance.com>

### CODE REVIEW

**Comment by pizhenwei** [comment]
@Yolandajn Hi, I always got zero PPS/BFS of **rx queue 0** on my enviroment (Mellanox Technologies MT27800 Family [ConnectX-5]).
And I checked mlnx driver code and TAP driver code, I guess we need to get queue id by kernel API `skb_get_rx_queue`. Then netqtop could show the right result. What do you think about this?

**Comment by Yolandajn** [comment]
> @Yolandajn Hi, I always got zero PPS/BFS of **rx queue 0** on my enviroment (Mellanox Technologies MT27800 Family [ConnectX-5]).
> And I checked mlnx driver code and TAP driver code, I guess we need to get queue id by kernel API `skb_get_rx_queue`. Then netqtop could show the right result. What do you think about this?

hi, thank you for the advice. The new version also works fine on my machine, so i think it is a good idea to make this change. BTW what did you get by skb->queue_mapping? was it `-1` or something else?
Besides, i was thinking that we could also change the probe for net_dev_start_xmit to avoid potential problems. But using `skb_get_tx_queue` caused compiling errors. Im not sure if it is necessary, or we should use other functions? If it's not necessay then we can just keep this version.

**Comment by pizhenwei** [comment]
> > @Yolandajn Hi, I always got zero PPS/BFS of **rx queue 0** on my enviroment (Mellanox Technologies MT27800 Family [ConnectX-5]).
> > And I checked mlnx driver code and TAP driver code, I guess we need to get queue id by kernel API `skb_get_rx_queue`. Then netqtop could show the right result. What do you think about this?
> 
> hi, thank you for the advice. The new version also works fine on my machine, so i think it is a good idea to make this change. BTW what did you get by skb->queue_mapping? was it `-1` or something else?
> Besides, i was thinking that we could also change the probe for net_dev_start_xmit to avoid potential problems. But using `skb_get_tx_queue` caused compiling errors. Im not sure if it is necessary, or we should use other functions? If it's not necessay then we can just keep this version.

It seems that you missed the change of this PR... Actually I'm trying to use `skb_get_tx_queue` to get the queue id.

I also tested for virtio-net case:
```
Linux-5.4
Virtio-net 4 queues
```
The original version netqop also got zero PPS/BFS of **rx queue 0**. The new version(with patch of this PR) worked fine.

And we can check [source code](https://github.com/torvalds/linux/blob/master/drivers/net/virtio_net.c) of function **static void receive_buf(...)** which calls **skb_record_rx_queue(...)** to mark rx queue id.

**Comment by pizhenwei** [comment]
@yonghong-song hi, what do you think about this change?

File: tools/netqtop.c [hunk 1]
```diff
@@ -95,19 +95,24 @@ TRACEPOINT_PROBE(net, net_dev_start_xmit){
 }
 
 TRACEPOINT_PROBE(net, netif_receive_skb){
-    struct sk_buff* skb = (struct sk_buff*)args->skbaddr;
-    if(!name_filter(skb)){
+    struct sk_buff skb;
+
+    bpf_probe_read(&skb, sizeof(skb), args->skbaddr);
```
Comment (yonghong-song) [inline tools/netqtop.c:100]: Is this change necessary? We can directly directly get fields from skb pointer. What is the need to copy skb structure to the stack? Did you hit any issues?

File: tools/netqtop.c [hunk 2]
```diff
@@ -95,19 +95,24 @@ TRACEPOINT_PROBE(net, net_dev_start_xmit){
...
+    if(!name_filter(&skb)){
         return 0;
     }
 
-    u16 qid = skb->queue_mapping;
+    u16 qid = 0;
+    if (skb_rx_queue_recorded(&skb))
+        qid = skb_get_rx_queue(&skb);
```
Comment (yonghong-song) [inline tools/netqtop.c:107]: From my understanding, 
```
static inline u16 skb_get_rx_queue(const struct sk_buff *skb)
{
        return skb->queue_mapping - 1;
} 
```

Here qid = 0 could mean two things: there is no mapping or actual qid of 0. Do we need to differentiate this? For example, if `!skb_rx_queue_recorded(&skb)`, simply return from the future as it is not in any queue? Or if not in any queue, it means queue 0?

**Review by yonghong-song — COMMENTED** [review]
sorry for late reply. In vacation and also with Christmas holiday last week...

**Comment by yonghong-song** [comment]
[buildbot, ok to test]

File: tools/netqtop.c (hunk 1 above)
Comment (pizhenwei) [reply tools/netqtop.c:100]: Actually we directly can get `queue_mapping` field from from skb pointer, but we can't directly call function `skb_rx_queue_recorded` & `skb_get_rx_queue` because bcc reports error. So two options:

- 1, re-implement `skb_rx_queue_recorded` & `skb_get_rx_queue` in this program.
- 2, copy skb and use kernel implemented function.

Solution 1 is not a hard work, but from my understanding, keeping the same API with kernel is better.

File: tools/netqtop.c (hunk 2 above)
Comment (pizhenwei) [reply tools/netqtop.c:107]: - If the NIC does **not** support multi-queue feature, there is only one queue(qid is always 0).
- And if the NIC supports multi-queue feature, there are several queues with different qid(from 0 to n-1).

So I guess current logic could represent the two cases. And if we want to know weather a NIC supports multi-queue feature, we can use `ethtool -l NICNAME` to test it.

File: tools/netqtop.c (hunk 1 above)
Comment (yonghong-song) [reply tools/netqtop.c:100]: Okay, I see. Let us use option 2 then. Re-implementing is not desirable.

File: tools/netqtop.c (hunk 2 above)
Comment (yonghong-song) [reply tools/netqtop.c:107]: What is the return value for skb_rx_queue_recorded(&skb) for single-queue feature? If this is false, then current code should work fine. Just adding some comments to explain is good enough.

Comment (pizhenwei) [reply tools/netqtop.c:116]: OK, added some comments and force-pushed!

**Comment by yonghong-song** [comment]
LGTM. Thanks!
